File: src/main/java/net/tlalka/imager/core/FileReader.java
package net.tlalka.imager.core;

import io.reactivex.Observable;

import javax.activation.MimetypesFileTypeMap;
import java.io.File;

public class FileReader {

    private static final File[] EMPTY_ARRAY = {};

    public Observable<File> getImages(String directory) {
        return Observable
                .fromArray(getFileListFromDirectory(directory))
                .filter(this::isImageFile);
    }

    private File[] getFileListFromDirectory(String path) {
        File[] files = new File(path).listFiles();
        return files != null ? files : EMPTY_ARRAY;
    }

    private boolean isImageFile(File file) {
        return new MimetypesFileTypeMap().getContentType(file).startsWith("image");
    }
}

File: src/main/kotlin/net/tlalka/imager/core/FileReader.kt
package net.tlalka.imager.core

import io.reactivex.Observable
import java.io.File
import javax.activation.MimetypesFileTypeMap

class FileReader {

    fun getImages(directory: String): Observable<File> {
        return Observable
                .fromArray(*getFileListFromDirectory(directory))
                .filter { this.isImageFile(it) }
    }

    private fun getFileListFromDirectory(path: String): Array<File> {
        return File(path).listFiles() ?: arrayOf<File>()
    }

    private fun isImageFile(file: File): Boolean {
        return MimetypesFileTypeMap().getContentType(file).startsWith("image", true)
    }
}
